### PR TITLE
[layer] Support in-place batch normalization

### PR DESCRIPTION
**Pull request by kparichay**
Support in-place batch normalization where the batch normalization
input/output is not stored and is over-written by the next layer.

This patch removes the input/output memory requirement when using
batch normalization layer.

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [x]Passed [ ]Failed [ ]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

### CODE REVIEW

**Comment by kparichay** [comment]
The reason this works is that batch-normalization stores a processed version of input and does not need storing the input.

File: nntrainer/models/neuralnet.cpp [hunk 1]
```diff
@@ -164,6 +164,29 @@ int NeuralNetwork::compile() {
...
+      /** @note assumes BatchNormalizationLayer is only for single in/out tensor
+       */
+      auto &prev_layer = model_graph.getLayerNode(l->input_layers[0]).layer;
+
+      unsigned int loc;
+      auto layer_name = l->getName();
+      for (loc = 0; loc < prev_layer->output_layers.size(); ++loc)
+        if (prev_layer->output_layers[loc] == layer_name)
+          break;
+
```
Comment (zhoonit) [inline nntrainer/models/neuralnet.cpp:182]: How about adding a check `if loc == prev_layer.output_layers.size()`
Otherwise net_hidden[loc] can overflow?

File: nntrainer/models/neuralnet.cpp [hunk 2]
```diff
@@ -164,6 +164,29 @@ int NeuralNetwork::compile() {
...
 
+void NeuralNetwork::inPlaceBatchNormOptimization() {
+  auto &sorted = model_graph.getSorted();
+
+  for (unsigned int idx = 1; idx < sorted.size() - 1; ++idx) {
+    auto &l = sorted[idx].layer;
+    if (l->getType() == BatchNormalizationLayer::type) {
+      /** @note assumes BatchNormalizationLayer is only for single in/out tensor
+       */
+      auto &prev_layer = model_graph.getLayerNode(l->input_layers[0]).layer;
```
Comment (zhoonit) [inline nntrainer/models/neuralnet.cpp:175]: Assure `l->input_layers.size() == 1`?

File: nntrainer/models/neuralnet.h [hunk 3]
```diff
@@ -100,7 +100,8 @@ class NeuralNetwork : public ml::train::Model {
     compiled(false),
     def_name_count(0),
     loadedFromConfig(false),
-    app_context(app_context_) {}
+    app_context(app_context_),
+    in_place_bn_layer_optimization(true) {}
```
Comment (zhoonit) [inline nntrainer/models/neuralnet.h:104]: How about we add a default argument for the option?

**Review by zhoonit — COMMENTED** [review]
as far as I understood, this optimization is implementation specific. If it is, could you add a note about how this works? My concern is that this might not work for other accelerated batch norm.

**Review by zhoonit — APPROVED** [review]
LGTM with minor comments

File: nntrainer/models/neuralnet.cpp (hunk 1 above)
Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:185]: added a throw.

File: nntrainer/models/neuralnet.cpp (hunk 2 above)
Comment (kparichay) [reply nntrainer/models/neuralnet.cpp:178]: added a throw.

File: nntrainer/models/neuralnet.h (hunk 3 above)
Comment (kparichay) [reply nntrainer/models/neuralnet.h]: added.

**Comment by kparichay** [comment]
> as far as I understood, this optimization is implementation specific. If it is, could you add a note about how this works? My concern is that this might not work for other accelerated batch norm.

Added a note in the optimization function declaration.

**Review by zhoonit — APPROVED** [review]
LGTM

**Review by jijoongmoon — APPROVED** [review]
LGTM
